### ISSUE TITLE
Support for conflict clause and with queries

### ISSUE THREAD

**Issue by 0xjac** [go-jet/jet#30]
Jet is awesome, but features are a bit limiting for my use cases. Any plan/timeline to add the following features (for postgres, I assume MySQL has an equivalent):

1. [Conflict clause](https://www.postgresql.org/docs/12/sql-insert.html#SQL-ON-CONFLICT) such as: 
    ``` sql
    INSERT INTO table (col) VALUES (val) ON CONFLICT DO NOTHING
    ```
2. [WITH Queries (CTE)](https://www.postgresql.org/docs/9.1/queries-with.html) such as:
    ``` sql
    WITH alias AS (
        SELECT x FROM y
    )
    SELECT * FROM z
    WHERE a IN (SELECT a FROM alias)
    ```

Thanks for the great work!

**Comment by go-jet** [comment]
Hi @0xjac. There is a plan to add both of these features. They probably will not be added in the next release(which will be released in a week or two), but with the release after. Rough estimate in a month or two.  

Regarding CTE is there a reason you can't use subqueries? Subqueries in jet resembles to CTE, and performance of subqueries and CTE are similar in postgres .  
Sample usage can be seen here: 
https://github.com/go-jet/jet/wiki/SELECT#sub-queries

**Comment by 0xjac** [comment]
Hi @go-jet, thanks for the very quick reply. I am happy to hear you plan on adding these in such a short time frame. It's great news for me, especially for the conflict clause.

Regarding the CTE, the main reason I am using them is because I have a raw SQL using a CTE which I am converting to jet and I did straight up without thinking first :wink: A subquery works just fine in my case. Thanks a lot for the reminder.

**Comment by 0xjac** [comment]
If anyone stumble upon this issue, I have used the following quick fix to solve it for now.
It's not the cleanest but it work in my case.

``` go
func OnConflictDoNothingWrapper(stmt InsertStatement) (query string, args []interface{}) {
	query, args = stmt.Sql()
	query = fmt.Sprintf("%s ON CONFLICT DO NOTHING;", query[:strings.LastIndex(query, ";")])
	return
}

func DebugOnConflictDoNothingWrapper(stmt InsertStatement) (query string) {
	query = stmt.DebugSql()
	query = fmt.Sprintf("%s ON CONFLICT DO NOTHING;", query[:strings.LastIndex(query, ";")])
	return
}
```

Then for any insert statement (`insertStmt`), you can use it as follow:

``` go
query, args := OnConflictDoNothingWrapper(insertStmt) // replaces insertStmt.Sql()
query := DebugOnConflictDoNothingWrapper(insertStmt) // replaces insertStmt.DebugSql()

**Comment by go-jet** [comment]
@0xjac There is now full `ON CONFLICT` support for postgres on `develop` branch.   
You can give it a try, before the official release. There might be some additional changes before release, but it is unlikely to be a breaking change.

Sample usege:

```go
stmt := Employee.INSERT(Employee.AllColumns).
	MODEL(employee).
	ON_CONFLICT(Employee.EmployeeID).DO_NOTHING()
```

or

```go
stmt := Link.INSERT(Link.ID, Link.URL, Link.Name, Link.Description).
	VALUES(100, "http://www.postgresqltutorial.com", "PostgreSQL Tutorial", DEFAULT).
	ON_CONFLICT(Link.ID).DO_UPDATE(
	        SET(Link.ID, Link.EXCLUDED.ID).
		SET(Link.URL, "http://www.postgresqltutorial2.com"),
        ).
	RETURNING(Link.AllColumns)
```

**Comment by Allezxandre** [comment]
Awesome! I had this issue as well, but I just tried with your code on `develop` and it works great. 
Thank you!

**Comment by 0xjac** [comment]
Thanks a lot I'll try it out. Feel free to close the issue

**Comment by go-jet** [comment]
Conflict clause and WITH statements are now included in the latest 2.3.0 release.

**Comment by BillBuilt** [comment]
Hello,

We just ran into a problem with ON_CONFLICT().DO_NOTHING().

it seems that the ON CONFLICT clause is omitted entirely from the SQL when ON_CONFLICT().DO_NOTHING() does not include a conflict target, however ON_CONFLICT().DO_NOTHING() is supposed to be able to accept no conflict target, only  DO UPDATE requires that, according to [Postgres](https://www.postgresql.org/docs/15/sql-insert.html#SQL-ON-CONFLICT).


**conflict_target**
Specifies which conflicts ON CONFLICT takes the alternative action on by choosing arbiter indexes. Either performs unique index inference, or names a constraint explicitly. **ior ON CONFLICT DO NOTHING, it is optional to specify a conflict_target; when omitted, conflicts with all usable constraints (and unique indexes) are handled**. For ON CONFLICT DO UPDATE, a conflict_target must be provided.
